### ISSUE TITLE
TestMachinesStoragePools.testStoragePoolsDeletion fails in TF on Fedora-38/39/rawhide?

### ISSUE THREAD

**Issue by jelly** [cockpit-project/cockpit-machines#1284]
It cannot reproduce this locally, so this must be something with either testing farm's repo's
```
> warning: GET_STORAGE_POOL action failed for path /org/libvirt/QEMU/storagepool/_ccca9334_8442_4d1d_a579_69748eecb26b Storage pool not found: no storage pool with matching uuid 'ccca9334-8442-4d1d-a579-69748eecb26b'
> warning: One of aria-label or aria-labelledby properties should be passed when using the progress component without a title.
> warning: One of aria-label or aria-labelledby properties should be passed when using the progress component without a title.
> warning: One of aria-label or aria-labelledby properties should be passed when using the progress component without a title.
> warning: One of aria-label or aria-labelledby properties should be passed when using the progress component without a title.
> warning: One of aria-label or aria-labelledby properties should be passed when using the progress component without a title.
> warning: One of aria-label or aria-labelledby properties should be passed when using the progress component without a title.
> warning: One of aria-label or aria-labelledby properties should be passed when using the progress component without a title.
> warning: One of aria-label or aria-labelledby properties should be passed when using the progress component without a title.
> warning: GET_STORAGE_VOLUMES action failed for pool iscsi-pool-3 : Storage pool not found: no storage pool with matching uuid '417d0f2e-edbf-4e56-90bb-51c224672e3a'
> warning: One of aria-label or aria-labelledby properties should be passed when using the progress component without a title.
> warning: One of aria-label or aria-labelledby properties should be passed when using the progress component without a title.
Wrote subVmTest1 XML to TestMachinesStoragePools-testStoragePoolsDeletion-fedora-38-localhost-22-subVmTest1.xml
Wrote subVmTest1 log to TestMachinesStoragePools-testStoragePoolsDeletion-fedora-38-localhost-22-subVmTest1.log
Traceback (most recent call last):
  File "/var/ARTIFACTS/work-storage1zzv8_0r/plans/all/storage/discover/default-0/tests/test/check-machines-storage-pools", line 1038, in testStoragePoolsDeletion
    activate=True).execute()
                   ^^^^^^^^^
  File "/var/ARTIFACTS/work-storage1zzv8_0r/plans/all/storage/discover/default-0/tests/test/check-machines-storage-pools", line 848, in execute
    self.verify()
  File "/var/ARTIFACTS/work-storage1zzv8_0r/plans/all/storage/discover/default-0/tests/test/check-machines-storage-pools", line 943, in verify
    self.test_obj.waitPoolRow(self.name, self.connection, present=False)
  File "/var/ARTIFACTS/work-storage1zzv8_0r/plans/all/storage/discover/default-0/tests/test/machineslib.py", line 108, in waitPoolRow
    b.wait_not_present(pool_row)
  File "/var/ARTIFACTS/work-storage1zzv8_0r/plans/all/storage/discover/default-0/tests/test/common/testlib.py", line 612, in wait_not_present
    self.wait_js_func('!ph_is_present', selector)
  File "/var/ARTIFACTS/work-storage1zzv8_0r/plans/all/storage/discover/default-0/tests/test/common/testlib.py", line 603, in wait_js_func
    self.wait_js_cond("%s(%s)" % (func, ','.join(map(jsquote, args))))
  File "/var/ARTIFACTS/work-storage1zzv8_0r/plans/all/storage/discover/default-0/tests/test/common/testlib.py", line 588, in wait_js_cond
    self.raise_cdp_exception("timeout\nwait_js_cond", cond, result["exceptionDetails"], trailer)
  File "/var/ARTIFACTS/work-storage1zzv8_0r/plans/all/storage/discover/default-0/tests/test/common/testlib.py", line 292, in raise_cdp_exception
    raise Error("%s(%s): %s" % (func, arg, msg))
testlib.Error: timeout
wait_js_cond(!ph_is_present("tbody tr[data-row-id=\"pool-iscsi-pool-3-system\"]")): condition did not become true
```

https://artifacts.dev.testing-farm.io/71dae27d-5555-4a74-b774-9c7d3303c0cc/

https://github.com/cockpit-project/cockpit-machines/pull/1282

Packages on TF:
```
+ grep -E 'libvirt|qemu'
ipxe-roms-qemu-20220210-3.git64113751.fc38.noarch
libvirt-client-9.0.0-3.fc38.x86_64
libvirt-daemon-9.0.0-3.fc38.x86_64
libvirt-daemon-config-network-9.0.0-3.fc38.x86_64
libvirt-daemon-driver-interface-9.0.0-3.fc38.x86_64
libvirt-daemon-driver-network-9.0.0-3.fc38.x86_64
libvirt-daemon-driver-nodedev-9.0.0-3.fc38.x86_64
libvirt-daemon-driver-qemu-9.0.0-3.fc38.x86_64
libvirt-daemon-driver-storage-core-9.0.0-3.fc38.x86_64
libvirt-daemon-driver-storage-disk-9.0.0-3.fc38.x86_64
libvirt-daemon-driver-storage-iscsi-9.0.0-3.fc38.x86_64
libvirt-daemon-driver-storage-iscsi-direct-9.0.0-3.fc38.x86_64
libvirt-daemon-driver-storage-logical-9.0.0-3.fc38.x86_64
libvirt-dbus-1.4.0-7.fc38.x86_64
libvirt-glib-4.0.0-8.fc38.x86_64
libvirt-libs-9.0.0-3.fc38.x86_64
python3-libvirt-9.0.0-2.fc38.x86_64
qemu-audio-spice-7.2.6-1.fc38.x86_64
qemu-block-curl-7.2.6-1.fc38.x86_64
qemu-char-spice-7.2.6-1.fc38.x86_64
qemu-common-7.2.6-1.fc38.x86_64
qemu-device-usb-host-7.2.6-1.fc38.x86_64
qemu-device-usb-redirect-7.2.6-1.fc38.x86_64
qemu-guest-agent-7.2.6-1.fc38.x86_64
qemu-img-7.2.6-1.fc38.x86_64
qemu-kvm-core-7.2.6-1.fc38.x86_64
qemu-system-x86-core-7.2.6-1.fc38.x86_64
qemu-ui-opengl-7.2.6-1.fc38.x86_64
qemu-ui-spice-core-7.2.6-1.fc38.x86_64
```

Packages on local F38 vm with updates-testnig
```
ipxe-roms-qemu-20220210-3.git64113751.fc38.noarch
libvirt-client-9.0.0-3.fc38.x86_64
libvirt-daemon-9.0.0-3.fc38.x86_64
libvirt-daemon-config-network-9.0.0-3.fc38.x86_64
libvirt-daemon-driver-interface-9.0.0-3.fc38.x86_64
libvirt-daemon-driver-network-9.0.0-3.fc38.x86_64
libvirt-daemon-driver-nodedev-9.0.0-3.fc38.x86_64
libvirt-daemon-driver-qemu-9.0.0-3.fc38.x86_64
libvirt-daemon-driver-storage-core-9.0.0-3.fc38.x86_64
libvirt-daemon-driver-storage-disk-9.0.0-3.fc38.x86_64
libvirt-daemon-driver-storage-iscsi-9.0.0-3.fc38.x86_64
libvirt-daemon-driver-storage-iscsi-direct-9.0.0-3.fc38.x86_64
libvirt-daemon-driver-storage-logical-9.0.0-3.fc38.x86_64
libvirt-dbus-1.4.1-1.fc38.x86_64
libvirt-glib-4.0.0-8.fc38.x86_64
libvirt-libs-9.0.0-3.fc38.x86_64
python3-libvirt-9.0.0-2.fc38.x86_64
qemu-audio-spice-7.2.6-1.fc38.x86_64
qemu-block-curl-7.2.6-1.fc38.x86_64
qemu-char-spice-7.2.6-1.fc38.x86_64
qemu-common-7.2.6-1.fc38.x86_64
qemu-device-usb-host-7.2.6-1.fc38.x86_64
qemu-device-usb-redirect-7.2.6-1.fc38.x86_64
qemu-guest-agent-7.2.6-1.fc38.x86_64
qemu-img-7.2.6-1.fc38.x86_64
qemu-kvm-core-7.2.6-1.fc38.x86_64
qemu-system-x86-core-7.2.6-1.fc38.x86_64
qemu-ui-opengl-7.2.6-1.fc38.x86_64
qemu-ui-spice-core-7.2.6-1.fc38.x86_64
qemu-user-static-7.2.6-1.fc38.x86_64
qemu-user-static-aarch64-7.2.6-1.fc38.x86_64
qemu-user-static-alpha-7.2.6-1.fc38.x86_64
qemu-user-static-arm-7.2.6-1.fc38.x86_64
qemu-user-static-cris-7.2.6-1.fc38.x86_64
qemu-user-static-hexagon-7.2.6-1.fc38.x86_64
qemu-user-static-hppa-7.2.6-1.fc38.x86_64
qemu-user-static-loongarch64-7.2.6-1.fc38.x86_64
qemu-user-static-m68k-7.2.6-1.fc38.x86_64
qemu-user-static-microblaze-7.2.6-1.fc38.x86_64
qemu-user-static-mips-7.2.6-1.fc38.x86_64
qemu-user-static-nios2-7.2.6-1.fc38.x86_64
qemu-user-static-or1k-7.2.6-1.fc38.x86_64
qemu-user-static-ppc-7.2.6-1.fc38.x86_64
qemu-user-static-riscv-7.2.6-1.fc38.x86_64
qemu-user-static-s390x-7.2.6-1.fc38.x86_64
qemu-user-static-sh4-7.2.6-1.fc38.x86_64
qemu-user-static-sparc-7.2.6-1.fc38.x86_64
qemu-user-static-x86-7.2.6-1.fc38.x86_64
qemu-user-static-xtensa-7.2.6-1.fc38.x86_64
qemu-virtiofsd-7.2.6-1.fc38.x86_64
```

**Comment by KKoukiou** [comment]
Can you paste the journal logs link here? I can check the libvirt logs there for this. @jelly

**Comment by jelly** [comment]
@KKoukiou that would be https://artifacts.dev.testing-farm.io/71dae27d-5555-4a74-b774-9c7d3303c0cc/work-storage1zzv8_0r/plans/all/storage/execute/data/guest/default-0/test/browser/storage-1/data/TestMachinesStoragePools-testStoragePoolsDeletion-fedora-38-localhost-22-FAIL.log.gz

**Comment by jelly** [comment]
Created https://github.com/cockpit-project/cockpit-machines/pull/1285 for debugging with extra debug output.

**Comment by jelly** [comment]
It succeed there, so I suppose this is a flake.

**Comment by martinpitt** [comment]
This happens in https://src.fedoraproject.org/rpms/cockpit-machines/pull-request/126 and https://src.fedoraproject.org/rpms/cockpit-machines/pull-request/124 , logs are linked from there.

**Comment by martinpitt** [comment]
It also happened in #1286 on [c8s](https://artifacts.dev.testing-farm.io/63e843a8-158c-4a12-9c4d-fac1e3051600/) and [rawhide](https://artifacts.dev.testing-farm.io/6754d92b-50eb-484e-b87f-3fd5eadc8e4c/).

**Comment by martinpitt** [comment]
The [journal](https://artifacts.dev.testing-farm.io/da04dd74-624c-4a47-8942-c9008254c0be/work-storagep49j0oi0/plans/all/storage/execute/data/guest/default-0/test/browser/storage-1/data/TestMachinesStoragePools-testStoragePoolsDeletion-fedora-40-localhost-22-FAIL.log.gz) has some very interesting bits:
```
systemd[1]: virtstoraged.service: Got notification message from PID 22047, but reception only permitted for main PID 19509
```
This doesn't happen on a local run.

```
AVC avc:  denied  { sys_rawio } for  pid=22048 comm="scsi_id" capability=17  scontext=system_u:system_r:virtstoraged_t:s0 tcontext=system_u:system_r:virtstoraged_t:s0 tclass=capability permissive=1
```
However, it is in permissive mode (ugh, why? our tests don't disable SELinux). On a local run this does not happen.
```
virtstoraged[19509]: this function is not supported by the connection driver: storage pool does not support vol deletion
virtqemud[19173]: this function is not supported by the connection driver: storage pool does not support vol deletion
```
This also happens in a local run, so I suppose the test covers an "undeletable" volume.

Doing another run in #1285

**Comment by martinpitt** [comment]
Collecting info here:
 - [Adding a `print(virsh pool-list)`](https://github.com/cockpit-project/cockpit-machines/commit/0864954515e9c5d224ceb1efdc6208a4a92dfe21#diff-61bd81ce3bad8562580c542c237ebf715677a027fe75b09f37590423a29a0dc2R887) reliably "fixes" the failure; it seems to trigger a re-scan or event in libvirt that papers over it.
 - Simply [waiting longer](https://github.com/cockpit-project/cockpit-machines/commit/69820b34ffaca67128103886b558a915dd8a2fbc) does not help.


I added some debug logging to the storage pool events. In a successful (local) run, this happens (empty lines are mine, for separating the phases):
```
+ virsh pool-define-as iscsi-pool-3 --type iscsi --target /dev/disk/by-id/ --source-host 127.0.0.1 --source-dev iqn.2019-09.cockpit.lan
> debug: StoragePoolEvent on system /org/libvirt/QEMU/storagepool/_2d7609d6_fbf0_47ef_b948_2cb884f0346c: DEFINED|CREATED
> debug: reducer vms: action={"type":"UPDATE_ADD_STORAGE_POOL","payload":{"storagePool":{"connectionName":"system","type":"iscsi","name":"iscsi-pool-3","id":"/org/libvirt/QEMU/storagepool/_2d7609d6_fbf0_47ef_b948_2cb884f0346c","uuid":"2d7609d6-fbf0-47ef-b948-2cb884f0346c","capacity":"0","available":"0","allocation":"0","target":{"path":"/dev/disk/by-id"},"source":{"host":{"name":"127.0.0.1"},"device":{"path":"iqn.2019-09.cockpit.lan"}},"active":false,"persistent":true,"autostart":false,"volumes":[]}}}

+ virsh pool-start iscsi-pool-3
> debug: StoragePoolEvent on system /org/libvirt/QEMU/storagepool/_2d7609d6_fbf0_47ef_b948_2cb884f0346c: STARTED
-> wait: ph_text_is("#pool-iscsi-pool-3-system-state","active")
> debug: reducer vms: action={"type":"UPDATE_ADD_STORAGE_POOL","payload":{"storagePool":{"connectionName":"system","type":"iscsi","name":"iscsi-pool-3","id":"/org/libvirt/QEMU/storagepool/_2d7609d6_fbf0_47ef_b948_2cb884f0346c","uuid":"2d7609d6-fbf0-47ef-b948-2cb884f0346c","capacity":"52428800","available":"0","allocation":"52428800","target":{"path":"/dev/disk/by-id"},"source":{"host":{"name":"127.0.0.1"},"device":{"path":"iqn.2019-09.cockpit.lan"}},"active":true,"persistent":true,"autostart":false,"volumes":[{"connectionName":"system","name":"unit:0:0:0","type":"block","path":"/dev/disk/by-id/scsi-36001405eb72025f2b1d4e4d9dd1232c6","capacity":"52428800","allocation":"52428800","physical":"52428800"}]},"updateOnly":true}}

+ virsh pool-refresh iscsi-pool-3; virsh vol-list iscsi-pool-3
> debug: StoragePool.Refresh on system
> debug: reducer vms: action={"type":"UPDATE_ADD_STORAGE_POOL","payload":{"storagePool":{"connectionName":"system","type":"iscsi","name":"iscsi-pool-3","id":"/org/libvirt/QEMU/storagepool/_2d7609d6_fbf0_47ef_b948_2cb884f0346c","uuid":"2d7609d6-fbf0-47ef-b948-2cb884f0346c","capacity":"52428800","available":"0","allocation":"52428800","target":{"path":"/dev/disk/by-id"},"source":{"host":{"name":"127.0.0.1"},"device":{"path":"iqn.2019-09.cockpit.lan"}},"active":true,"persistent":true,"autostart":false,"volumes":[{"connectionName":"system","name":"unit:0:0:0","type":"block","path":"/dev/disk/by-id/scsi-36001405eb72025f2b1d4e4d9dd1232c6","capacity":"52428800","allocation":"52428800","physical":"52428800"}]}}}

-> wait: !ph_is_present("div.pf-v5-c-modal-box")
> debug: StoragePoolEvent on system /org/libvirt/QEMU/storagepool/_2d7609d6_fbf0_47ef_b948_2cb884f0346c: STOPPED
> debug: StoragePoolEvent on system /org/libvirt/QEMU/storagepool/_2d7609d6_fbf0_47ef_b948_2cb884f0346c: UNDEFINED
> debug: reducer vms: action={"type":"UNDEFINE_STORAGE_POOL","payload":{"connectionName":"system","id":"/org/libvirt/QEMU/storagepool/_2d7609d6_fbf0_47ef_b948_2cb884f0346c"}}
> debug: reducer vms: action={"type":"UNDEFINE_STORAGE_POOL","payload":{"connectionName":"system","id":"/org/libvirt/QEMU/storagepool/_2d7609d6_fbf0_47ef_b948_2cb884f0346c"}}
-> wait: !ph_is_present("tbody tr[data-row-id=\"pool-iscsi-pool-3-system\"]")
```

In a [failing test](https://artifacts.dev.testing-farm.io/61522869-1506-4339-bcf7-58591ccbd878/) it looks like this:
```
+ virsh pool-define-as iscsi-pool-3 --type iscsi --target /dev/disk/by-id/ --source-host 127.0.0.1 --source-dev iqn.2019-09.cockpit.lan
> debug: StoragePoolEvent on system /org/libvirt/QEMU/storagepool/_3b041ff1_d61e_40c1_af2d_b6940b59d61e: DEFINED|CREATED
+ virsh pool-list --all
> debug: reducer vms: action={"type":"UPDATE_ADD_STORAGE_POOL","payload":{"storagePool":{"connectionName":"system","type":"iscsi","name":"iscsi-pool-3","id":"/org/libvirt/QEMU/storagepool/_3b041ff1_d61e_40c1_af2d_b6940b59d61e","uuid":"3b041ff1-d61e-40c1-af2d-b6940b59d61e","capacity":"0","available":"0","allocation":"0","target":{"path":"/dev/disk/by-id"},"source":{"host":{"name":"127.0.0.1"},"device":{"path":"iqn.2019-09.cockpit.lan"}},"active":false,"persistent":true,"autostart":false,"volumes":[]}}}

+ virsh pool-start iscsi-pool-3
-> wait: ph_is_present("#pool-iscsi-pool-3-system-state")
> debug: StoragePoolEvent on system /org/libvirt/QEMU/storagepool/_3b041ff1_d61e_40c1_af2d_b6940b59d61e: STARTED
<- {'type': 'undefined'}
-> wait: ph_text_is("#pool-iscsi-pool-3-system-state","active")
> debug: reducer vms: action={"type":"UPDATE_ADD_STORAGE_POOL","payload":{"storagePool":{"connectionName":"system","type":"iscsi","name":"iscsi-pool-3","id":"/org/libvirt/QEMU/storagepool/_3b041ff1_d61e_40c1_af2d_b6940b59d61e","uuid":"3b041ff1-d61e-40c1-af2d-b6940b59d61e","capacity":"52428800","available":"0","allocation":"52428800","target":{"path":"/dev/disk/by-id"},"source":{"host":{"name":"127.0.0.1"},"device":{"path":"iqn.2019-09.cockpit.lan"}},"active":true,"persistent":true,"autostart":false,"volumes":[{"connectionName":"system","name":"unit:0:0:0","type":"block","path":"/dev/disk/by-id/wwn-0x600140570f3525795f14079aab21a401","capacity":"52428800","allocation":"52428800","physical":"52428800"}]},"updateOnly":true}}

+ virsh pool-refresh iscsi-pool-3; virsh vol-list iscsi-pool-3
> debug: StoragePool.Refresh on system
[... notably, NO UPDATE_ADD_STORAGE_POOL reducer! ... ]

-> wait: !ph_is_present("div.pf-v5-c-modal-box")
> debug: StoragePoolEvent on system /org/libvirt/QEMU/storagepool/_3b041ff1_d61e_40c1_af2d_b6940b59d61e: STOPPED
> debug: StoragePoolEvent on system /org/libvirt/QEMU/storagepool/_3b041ff1_d61e_40c1_af2d_b6940b59d61e: UNDEFINED
> debug: reducer vms: action={"type":"UNDEFINE_STORAGE_POOL","payload":{"connectionName":"system","id":"/org/libvirt/QEMU/storagepool/_3b041ff1_d61e_40c1_af2d_b6940b59d61e"}}
> warning: GET_STORAGE_VOLUMES action failed for pool iscsi-pool-3 : Requested operation is not valid: storage pool 'iscsi-pool-3' is not active
> debug: reducer vms: action={"type":"UPDATE_ADD_STORAGE_POOL","payload":{"storagePool":{"connectionName":"system","type":"iscsi","name":"iscsi-pool-3","id":"/org/libvirt/QEMU/storagepool/_3b041ff1_d61e_40c1_af2d_b6940b59d61e","uuid":"3b041ff1-d61e-40c1-af2d-b6940b59d61e","capacity":"52428800","available":"0","allocation":"52428800","target":{"path":"/dev/disk/by-id"},"source":{"host":{"name":"127.0.0.1"},"device":{"path":"iqn.2019-09.cockpit.lan"}},"active":true,"persistent":true,"autostart":false}}}
> warning: GET_STORAGE_POOL action failed for path /org/libvirt/QEMU/storagepool/_3b041ff1_d61e_40c1_af2d_b6940b59d61e Storage pool not found: no storage pool with matching uuid '3b041ff1-d61e-40c1-af2d-b6940b59d61e'
```

These two error messages point to a race condition.
